pkg/arvo/neo/cod/std/src/fil/messenger.md

```diff
@@ -0,0 +1,557 @@
+            [%make %groupchat `groupchat-diff/vax ~]
+        ==
+      ==
+    ::
+        %messenger-diff
+      ?>  =(our ship.src):bowl
+      =/  poke  !<(messenger-diff vax)
+      ?-    -.poke
+          %new-dm
+        =/  provider  ~[p/partner.poke %home %messenger]
+        :_  state
+        :~  :-  (welp here.bowl ~[%dms p/partner.poke])
+            [%make %dm `dm-diff/!>([%initiate partner.poke provider]) ~]
+        ==
+      ::
+          %new-groupchat
+        :_  state
+        :~  :-  (welp here.bowl ~[%groupchats p/our.bowl t/name.poke])
+            [%make %groupchat ~ ~]
+        ==
+      ::
+          %invite-to-groupchat
+        =/  provider  ~[p/ship.poke %home %messenger]
+        :_  state
+        :~  :-  (welp here.bowl ~[%groupchats p/our.bowl t/name.poke])
+            [%poke groupchat-diff/!>([%invite ship.poke provider])]
+        ==
+      ==
+    ==
+  --
+--
+```
+
+## The +deps arm
+The only part of this system that uses the `+deps` arm is `/message-sub`.
+
+```hoon
+::
+::  define dependencies
+++  deps
+  ::  deps:neo is (map term fief:neo)
+  %-  ~(gas by *deps:neo)
+  :~  :*  %pub
+          ::  fief:neo is [required=? =quay]
+          req=&
+          ::  quay:neo is (pair lash:neo (unit port:neo)
+          ::  lash:neo defines the shrub's state and pokes
+          [[%pro %sig] (sy %sig ~)]
+          %-  some
+          ::  port:neo constrains the shrub's kids
+          :-  %y
+          %-  ~(gas by *lads:neo)
+          :~  :-  [[%.n %da] |]
+              [[%pro %message] ~]
+          ==
+      ==
+  ==
+```
+
+With regards to the lifecycle of the shrub, the `+deps` arm types the shrubs whose names and locations are passed in as an argument in the `%make` card when this shrub is created.
+
+```hoon
+[%make %message-sub ~ (malt ~[[%pub (snoc host.poke %pub)]])]
+```
+
+In this card, `+malt` creates a `(map term pith:neo)` where the `term` is `%pub` and the path represented by the `pith` is `/path/to/chat/pub`. At `/path/to/chat/pub`, there’s a shrub which is the canonical “publisher” or “host” of a chat, whether that’s a group chat or a DM. Whenever there’s a state change in the publisher shrub, the `/message-sub` created in this card will be notified about it, but we’ll cover that in more detail in the next section.
+
+The `%pub` `term` will act as the key in the `map` of dependencies. This means `/message-sub` could have several dependencies. `%pub` doesn’t refer to anything, it’s just being defined in the `+deps` arm to differentiate between shrubs being passed in the card’s `conf:neo`; think of how you’d use head tags to route different cases in the `+on-poke` arm of a Gall agent.
+
+The `required` flag specfies whether the shrub being described in `+deps` needs to already exist in order for this shrub to build. If no publisher exists at `/path/to/chat/pub`, the `/message-sub` we’re trying to `%make` here will fail to build. It can only exist with reference to the publisher shrub.
```

Comment:
The `required` flag specifies that when this shrub is made, it must be passed in a `conf` that contains paths to existing shrubs. If we don't pass in a conf, or if no publisher exists at `/path/to/pub`, the `/message-sub` we’re trying to `%make` here will fail to build. It can only exist with reference to the publisher shrub.